### PR TITLE
Enabled Directory and File Creation for Files that Don't Exist at Runtime

### LINKED ISSUE #20: run_operations -- creating directories/files if not exist
It will be nice  if run_operation could generate all needed directories and files (including stored_dates) if they don't exist.  That would allow us to design a big, second unit test including all/most datasets.

I would remove the OPERATIONS directory and then run_operations would just work.

### PR DESCRIPTION
**Pull request by Ovec8hkin**
`run_operations` relies on a multitude of files and directories existing for the proper organization of log, error, and template files. Perviously if any of those directories or files didn't exist when accessed, run_operations would simply throw and error and exit. 

This Pull Request implements proper handling of file and directory creation so that, should those directories or files not exist, they are created, ensuring that manual creation of those files and folders doesn't need to be done before running `run_operations` on a new system.

Closes #20

### CODE REVIEW

File: run_operations.py [hunk 1]
```diff
@@ -119,168 +128,181 @@ def command_line_parse(args):
...
-	
-	with open('/nethome/'+user+'/insarlab/OPERATIONS/TEMPLATES/'+dataset+'.template', 'r') as template_file:
+	""" Reads template file for the current dataset and parses out the 'ssaraopt' option before creating command options
+		line options array for ssara_federated_query.py
+
+		Parameters: none
+		Returns:    ssara_options: [str], array of command line options to run ssara_federated_query with
+	"""
+
+	with open('/nethome/' + user + '/insarlab/OPERATIONS/TEMPLATES/' + dataset + '.template', 'r') as template_file:
```
Comment (gravelcycles) [inline run_operations.py:143]: Seems like an overly complicated and system-dependent method of finding the HOME path. See https://stackoverflow.com/a/4028943/8878070. Let's try to refactor out `user` in the process

File: run_operations.py [hunk 2]
```diff
@@ -35,41 +37,41 @@
...
 
-dataset = 'GalapagosSenDT128VV'		                				        # Single Dataset for Testing
+dataset = 'GalapagosSenDT128VV'  # Single Dataset for Testing
 
-user = subprocess.check_output(['whoami']).decode('utf-8').strip("\n")  			# Currently logged in user
-	
-stored_date = None			  							# previously stored date
-most_recent = None										# date parsed from SSARA
-inps = None;				       							# command line arguments
+user = subprocess.check_output(['whoami']).decode('utf-8').strip("\n")  # Currently logged in user
```
Comment (gravelcycles) [inline run_operations.py:42]: Should be removed after changing to automatic expansion of home directory

File: run_operations.py [hunk 3]
```diff
@@ -119,168 +128,181 @@ def command_line_parse(args):
...
+
 	most_recent_data = ssara_output.split("\n")[-2]
 	most_recent = datetime.strptime(most_recent_data.split(",")[3], date_format)
 
+	# Checks if the stored_date.date file exists and created the file+directories if it doesn't
+	filename = os.getenv('OPERATIONS') + '/stored_date.date'
+	if not os.path.exists(filename):
+		if not os.path.exists(os.path.dirname(filename)):
+			os.makedirs(os.path.dirname(filename))
+		subprocess.Popen("touch "+filename, shell=True).wait()
```
Comment (gravelcycles) [inline run_operations.py:175]: Seems like a code smell to create a new file using bash. Try: https://www.tutorialspoint.com/How-to-create-an-empty-file-using-Python

File: run_operations.py [hunk 4]
```diff
@@ -119,168 +128,181 @@ def command_line_parse(args):
...
+
 	# Write Most Recent Date to File
-	with open(os.getenv('OPERATIONS')+'/stored_date.date', 'rb') as stored_date_file:
-	
+	with open(filename, 'rb') as stored_date_file:
+
 		try:
-			date_line = subprocess.check_output(['grep', dataset, os.getenv('OPERATIONS')+'/stored_date.date']).decode('utf-8')
+			date_line = subprocess.check_output(
```
Comment (gravelcycles) [inline run_operations.py:181]: Can we do this with Python rather than bash?

File: run_operations.py [hunk 5]
```diff
@@ -119,168 +128,181 @@ def command_line_parse(args):
...
+				data = str(dataset + ": " + str(datetime.strftime(most_recent, date_format)) + "\n")
 				date_file.write(data)
 
-"""
-    Compares the most recent and stored dates 
-    Parameters: none
-    Returns:    boolean, whether the most recent data is more recent than the stored date
-"""
+
 def compare_dates():
```
Comment (gravelcycles) [inline run_operations.py:193]: Can you refactor out this function? It is only used in one spot and is just as readable if it were inline

File: run_operations.py [hunk 6]
```diff
@@ -332,31 +352,32 @@ def post_processing(files_to_move):
...
 			for i, file in enumerate(all_output_files):
 				if os.path.exists(file) and os.path.isfile(file):
-					files_to_move = [all_output_files[i], all_output_files[i+1]]
+					files_to_move = [all_output_files[i], all_output_files[i + 1]]
 					post_processing(files_to_move)
-				
-					all_output_files[:] = [f for fi in files_to_move if fi not in file and fi not in all_output_files[i+1]]
-				
+
+					all_output_files[:] = [f for fi in files_to_move if
```
Comment (gravelcycles) [inline run_operations.py:375]: What is this doing? Hard to understand, so a refactor might be in order.

**Review by gravelcycles — CHANGES_REQUESTED** [review]
The changes made for file creation are pretty straight forward and should be good. I made some comments on some of the underlying code that was committed before and could use a refactor. Not all are needed to be changed right now but please provide your thoughts.

Also, for a later diff, this file needs to be converted from using tabs to spaces. I know, I know it's funny. But all kidding aside, Python code gets weird at compile time if there are both tabs and spaces. The standard is spaces (usually 2 spaces, but I think RSMAS ISCE uses 4 spaces).

File: run_operations.py (hunk 6 above)
Comment (gravelcycles) [reply run_operations.py:375]: Something definitely to be done now is to change the variable name `file` to something else as `file` is a builtin

File: run_operations.py (hunk 1 above)
Comment (Ovec8hkin) [reply run_operations.py:143]: Yes. Falk has a better way using os.getenv(). Need to update that piece

File: run_operations.py (hunk 3 above)
Comment (Ovec8hkin) [reply run_operations.py:175]: Ive tried those, and the file never gets created. Im not happy with the way that works either, but the pythonic ways aren't working.

File: run_operations.py (hunk 4 above)
Comment (Ovec8hkin) [reply run_operations.py:181]: We could probably use `glob`

File: run_operations.py (hunk 6 above)
Comment (Ovec8hkin) [reply run_operations.py:375]: Yes should stop using file as variables name. This essentially the output file and the error file after they've been processed. Its really short hand way, but the expanded loop looks really messy as well.

Comment (Ovec8hkin) [reply run_operations.py:375]: If you have a better way to accomplish that, please comment

File: run_operations.py (hunk 4 above)
Comment (Ovec8hkin) [reply run_operations.py:181]: `glob` actually only works for gathering filenames from the filesystem not for file reading. We can do it in python my opening the file and reading each line until we find a desired match, but that doesn't seem to be any more elegant than this.

Comment (gravelcycles) [reply run_operations.py:181]: Would the `re` library work? Link to [stack-overflow discussion](https://stackoverflow.com/questions/1921894/grep-and-python)

File: run_operations.py (hunk 6 above)
Comment (gravelcycles) [reply run_operations.py:375]: Let's move this to a separate diff. I don't want to mess with this change, so I'll wait until this PR lands to `master`

**Pull request closed without merging**